### PR TITLE
fix: guard serial connect/disconnect race and wrong-port auto-select during flashing

### PR DESCRIPTION
**Pull request by nerdCopter**
AI Generated pull-request

## Summary

Firmware flashing could sometimes stop after the "Flashing" progress label appears, or fail to find the serial port after a reboot, with no consistent trigger. Root cause is the same class of defect as emuflight/EmuConfigurator#638 Findings 2 and 3 (originally filed against save-and-reboot tabs only), generalized to the firmware flasher's own reboot/reconnect path:

- `main.js`'s shared `_serialPort` handle had no generation tracking. A `serial-connect`/`serial-disconnect` racing across an `await` (the exact churn produced by `stm32.js`'s reboot-to-DFU handshake) could throw `Cannot read properties of null (reading 'on')` or attach listeners to a stale port instance. Fixed with a generation-token guard: each connect captures a token and re-checks it after every `await` before touching shared state.
- `port_handler.js`'s new-port auto-select picked `new_ports[0]` with no identity check, so an unrelated serial device (e.g. Bluetooth RFCOMM) enumerating in the same ~500ms poll window could be auto-selected ahead of the flight controller. Fixed by filtering known non-FC device paths and only acting on an unambiguous single candidate, re-evaluating on later polls otherwise.

Reviewed with local CodeRabbit CLI (3 rounds, 0 findings on final code -- round 1 caught a real ambiguous-candidate bug, fixed), and opencode (4 additional real issues found via full call-graph trace and fixed: `GUI.connecting_to` left stuck on a superseded connect, an unguarded double-close edge case, app-quit not invalidating an in-flight connect's generation, and the port-picker dropdown silently losing the user's manual selection during an ambiguous poll).

## Test plan
- [x] `node --check` on all modified files
- [x] `yarn lint` -- 0 errors (unchanged from baseline)
- [x] Headless `yarn dev` launch smoke-test (Xvfb) -- no JS-level crash
- [x] Hardware: PYRODRONEF7 flashed successfully 3x on `/dev/ttyACM1` with a phone connected as `/dev/ttyACM0` throughout (confirms normal flashing and the candidate-filtering fix both work correctly with another serial device present)
- [x] Hardware: confirmed operation-in-progress gating (`GUI.connect_lock`) correctly blocks interleaving actions during an active erase/flash
- [ ] Not tested: the exact concurrent-connect race requires switching away from the firmware_flasher tab mid-flash, which is gated off during normal single-tab use -- no test harness exists in this repo (see AGENTS.md) to verify this synthetically

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved connection handling during port opening, closing, and shutdown to prevent stale operations from affecting active connections.
  * Prevented failed or cancelled connection attempts from blocking future automatic connections.
  * Excluded Bluetooth RFCOMM ports from flight-controller detection.
  * Improved port selection when multiple possible flight-controller ports are detected.
  * Prevented stale data and error events from disconnected ports from being forwarded.
  * Preserved the current port selection when no unique flight-controller candidate is available.
  * Ensured delayed connections target the originally selected port.

### CODE REVIEW

File: main.js [hunk 1]
```diff
@@ -1351,6 +1373,7 @@ async function cleanupConnectionsBeforeQuit() {
   _usbOpenDevices.clear();
 
   if (_serialPort && _serialPort.isOpen) {
+    ++_serialGen; // invalidate any serial-connect still in flight during shutdown
```
[the remark this answers is not in the document]
Comment (nerdCopter) [reply main.js]: Fixed in e137ff28 -- moved `++_serialGen` before the `if (_serialPort && _serialPort.isOpen)` condition, per the suggested diff.

**Comment by nerdCopter** [comment]
@CodeRabbitAI, AI generated comment

Addressed the port_handler.js:141-170 outside-diff finding in e137ff28:

- Deferred auto-connect timeout: fixed. Captured `fc_candidates[0]` at schedule time (`detected_candidate`); the delayed callback now requires `selectedPort === detected_candidate` before clicking connect, instead of accepting any non-zero selection.
- Manual selection overwritten by an unambiguous candidate (line 144): not changed. Verified against `upstream/master` -- `$('div#port-picker #port').val(new_ports[0])` runs unconditionally whenever `!GUI.connected_to` and a new port appears, with no manual-selection check, in the pre-existing code. This branch didn't introduce or change that behavior; it's the same auto-select-over-manual precedent the app already had. Out of scope for this fix.

**Comment by nerdCopter** [comment]
@CodeRabbitAI, AI generated comment

Filed as #673.

**Comment by nerdCopter** [comment]
@CodeRabbitAI, AI generated comment

All findings from your review (inline: main.js gen-increment ordering; outside-diff: port_handler.js auto-connect binding) are addressed in e137ff28, and the Codacy strict-equality finding is addressed in 2a720c0a. Is anything still outstanding on this PR, or is it ready to merge from your side?
